lib/deepcrawl.js

```diff
@@ -0,0 +1,137 @@
+            .then(this.handleResponse);
+        }
+
+        // convert all camelCased options to _ for deepcrawl
+        const data = humps.decamelizeKeys(options);
+        return needle[`${method}Async`](url, data, requestOpts)
+          .then(this.handleResponse);
+      });
+  }
+
+  generateMethod (requiredFields, url, method) {
+    return (options) => {
+      options = options || {};
+      if (!options.sessionToken && !this.getSessionToken) {
+        throw new Error(`"sessionToken" or "getSessionToken" method required`);
+      }
+
+      requiredFields.forEach((field) => {
+        if (!options[field] && !this[field]) {
+          throw new Error(`${field} is required`);
+        }
+        // if the url contains {field}, replace {field} with the passed
+        // in value
+        if (url.indexOf(`{${field}}`) > -1) {
+          url = url.replace(`{${field}}`, (options[field] || this[field]));
+        }
+      });
+
+      return this.generateRequest(url, options, method);     
+    };
+  }
+
+  addMethodsToEndpoints (API, schema) {
+    for (const resource in API) {
+      const resourceUrl = this.baseUrl + schema.resources[resource].route,
+        requiredFields = _.clone(schema.resources[resource].requiredFields || []),
+        id = schema.resources[resource].id;
+
+      for (const action in schema.resources[resource].actions) {
+        let actionRequiredFields = requiredFields,
+          actionUrl = resourceUrl;
+        // check to see whether the resource action has extra data associated with it
+        const isString = typeof schema.resources[resource].actions[action] === 'string';
+        // the following actions should all have an "id" field associated with them
+        // we will append the id field to the required fields for the action, and
+        // add /{id} to the url, which will be replaced later with the passed in
+        // field
+        if (action === 'read' || action === 'update' || action === 'delete') {
+          actionRequiredFields = [id].concat(requiredFields);
+          actionUrl = resourceUrl + `/{${id}}`;
+        }
+        // the action may still have extra requiredFields needed to append
+        if (!isString && schema.resources[resource].actions[action].requiredFields) {
+          actionRequiredFields = actionRequiredFields.concat(schema.resources[resource].actions[action].requiredFields);
+        }
+        // determine the request method ('get', 'patch', 'delete', etc)
+        const method = (isString ? schema.resources[resource].actions[action] : schema.resources[resource].actions[action].type).toLowerCase();
+        API[resource][action] = this.generateMethod(actionRequiredFields, actionUrl, method);    
+      }
+    }
+  }
+
+  getAPI (opts) {
+    opts = opts || {};
+    this.accountId = opts.accountId || this.accountId;
+    // create the base API
+    const API = this.createEndpoints(schema);
+    this.addMethodsToEndpoints(API, schema);
```

Comment:
At the moment the `accountId` is cached and the API is dynamically created on each call to this method. Ultimately we'll probably want to reverse this -- cache the API after initial creation, and only create a new instance if a different version is requested (future). For now it should never change, and the only things that could change from one call to the next are the variables like account id, API key, etc. The basic schema should never change during the lifetime of this class instance.